signer/xpi/cose.go

```diff
@@ -0,0 +1,339 @@
+			err = errors.Wrapf(err, "failed to generate private key")
+			return
+		}
+		eeKey = signer.PrivateKey
+		eePublicKey = eeKey.(*ecdsa.PrivateKey).Public()
+	}
+	return
+}
+
+// isSupportedCOSEAlgValue returns whether the COSE alg value is supported or not
+func isSupportedCOSEAlgValue(algValue interface{}) bool {
+	return (algValue == cose.PS256.Value ||
+		algValue == cose.ES256.Value ||
+		algValue == cose.ES384.Value ||
+		algValue == cose.ES512.Value)
+}
+
+
+func expectHeadersAndGetKeyID(actual, expected *cose.Headers) (kidValue interface{}, err error) {
+	if actual == nil || expected == nil {
+		err = errors.New("xpi: cannot compare nil COSE headers")
+		return
+	}
+	if len(actual.Unprotected) != len(expected.Unprotected) {
+		err = fmt.Errorf("xpi: unexpected non-empty Unprotected headers got: %v", actual.Unprotected)
+		return
+	}
+	if len(actual.Protected) != len(expected.Protected) {
+		err = fmt.Errorf("xpi: unexpected Protected headers got: %v expected: %v", actual.Protected, expected.Protected)
+		return
+	}
+	if _, ok := expected.Protected[algHeaderValue]; ok {
+		algValue, ok := actual.Protected[algHeaderValue]
+		if !ok {
+			err = fmt.Errorf("xpi: missing expected alg in Protected Headers")
+			return
+		}
+		if !isSupportedCOSEAlgValue(algValue) {
+			err = fmt.Errorf("xpi: alg %v is not supported", algValue)
+			return
+		}
+	}
+	if _, ok := expected.Protected[kidHeaderValue]; ok {
+		kidValue, ok = actual.Protected[kidHeaderValue]
+		if !ok {
+			err = fmt.Errorf("xpi: missing expected kid in Protected Headers")
+			return
+		}
+	}
+	return
+}
+
+var (
+	expectedMessageHeaders = &cose.Headers{
+		Unprotected: map[interface{}]interface{}{},
+		Protected: map[interface{}]interface{}{
+			kidHeaderValue: nil,
+		},
+	}
+	expectedSignatureHeaders = &cose.Headers{
+		Unprotected: map[interface{}]interface{}{},
+		Protected: map[interface{}]interface{}{
+			kidHeaderValue: nil,
+			algHeaderValue: nil,
+		},
+	}
+)
+
+// isValidCOSESignature checks whether a COSE signature is valid for XPIs
+func isValidCOSESignature(sig *cose.Signature) (eeCert *x509.Certificate, err error) {
+	if sig == nil {
+		err = errors.New("xpi: cannot validate nil COSE Signature")
+		return
+	}
+
+	kidValue, err := expectHeadersAndGetKeyID(sig.Headers, expectedSignatureHeaders)
+	if err != nil {
+		err = errors.Wrapf(err, "xpi: got unexpected COSE Signature headers")
+		return
+	}
+
+	kidBytes, ok := kidValue.([]byte)
+	if !ok {
+		err = fmt.Errorf("xpi: COSE Signature kid value is not a byte array")
+		return
+	}
+
+	eeCert, err = x509.ParseCertificate(kidBytes)
+	if err != nil {
+		err = errors.Wrapf(err, "xpi: failed to parse X509 EE certificate from COSE Signature")
+		return
+	}
+
+	return
+}
+
+// isValidCOSEMessage checks whether a COSE SignMessage is a valid for
+// XPIs and returns parsed intermediate and end entity certs
+func isValidCOSEMessage(msg *cose.SignMessage) (intermediateCerts, eeCerts []*x509.Certificate, err error) {
```

Comment:
same as above, the naming isn't very clear. maybe `extractCertsFromCOSEMessage` ?